cpp/src/lance/arrow/updater.cc

```diff
@@ -0,0 +1,209 @@
+//  Copyright 2022 Lance Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+#include "lance/arrow/updater.h"
+
+#include <arrow/dataset/type_fwd.h>
+#include <arrow/result.h>
+#include <fmt/format.h>
+
+#include <filesystem>
+#include <limits>
+#include <memory>
+#include <vector>
+
+#include "lance/arrow/dataset.h"
+#include "lance/arrow/dataset_ext.h"
+#include "lance/arrow/file_lance.h"
+#include "lance/arrow/fragment.h"
+#include "lance/arrow/utils.h"
+#include "lance/format/data_fragment.h"
+#include "lance/format/schema.h"
+#include "lance/io/writer.h"
+
+namespace fs = std::filesystem;
+
+namespace lance::arrow {
+
+class Updater::Impl {
+ public:
+  Impl(std::shared_ptr<LanceDataset> dataset,
+       ::arrow::dataset::FragmentVector fragments,
+       std::shared_ptr<lance::format::Schema> full_schema,
+       std::shared_ptr<lance::format::Schema> column_schema)
+      : dataset_(std::move(dataset)),
+        full_schema_(std::move(full_schema)),
+        column_schema_(std::move(column_schema)),
+        fragments_(std::move(fragments)),
+        fragment_it_(fragments_.begin()) {}
+
+  /// Copy constructor
+  Impl(const Impl& other)
+      : dataset_(other.dataset_),
+        full_schema_(other.full_schema_),
+        column_schema_(other.column_schema_),
+        fragments_(other.fragments_.begin(), other.fragments_.end()),
+        fragment_it_(fragments_.begin()) {}
+
+  ::arrow::Result<std::shared_ptr<::arrow::RecordBatch>> Next();
+
+  ::arrow::Status Update(const std::shared_ptr<::arrow::Array>& arr);
+
+  ::arrow::Result<std::shared_ptr<LanceDataset>> Finish();
+
+ private:
+  auto data_dir() const { return dataset_->impl_->data_dir(); }
+
+  const auto& fs() const { return dataset_->impl_->fs; }
+
+  /// Prepare to read the next fragment.
+  ::arrow::Status NextFragment();
+
+  std::shared_ptr<LanceDataset> dataset_;
+  /// The schema of the newly generated dataset.
+  std::shared_ptr<lance::format::Schema> full_schema_;
+  /// The schema of the column to updated/added
+  std::shared_ptr<lance::format::Schema> column_schema_;
```

Comment:
so this can support adding multiple columns at a time?

Reply:
In theory, yes. 

This is internal API, we could change it to be `shared_ptr<lance::Field>` if we'd live to be more strict for now.